balancer/rls/internal/balancer_test.go

```diff
@@ -0,0 +1,812 @@
+	// server certificate used for the RLS server and the backend specifies a
+	// DNS SAN of "*.test.example.com". Hence we use a dial target which is a
+	// subdomain of the same here.
+	cc, err := grpc.Dial(r.Scheme()+":///rls.test.example.com", grpc.WithResolvers(r), grpc.WithTransportCredentials(clientCreds))
+	if err != nil {
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	defer cc.Close()
+
+	// Make an RPC and ensure it gets routed to the test backend.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	makeTestRPCAndExpectItToReachBackend(ctx, t, cc, backendCh)
+
+	// Make sure an RLS request is sent out.
+	verifyRLSRequest(t, rlsReqCh, true)
+
+	// Ensure a connection is established to the first RLS server.
+	if _, err := lis.newConnCh.Receive(ctx); err != nil {
+		t.Fatal("Timeout expired when waiting for LB policy to create control channel")
+	}
+}
+
+// TestConfigUpdate_DefaultTarget tests the scenario where a config update
+// changes the default target. Verifies that RPCs get routed to the new default
+// target after the config has been applied.
+func (s) TestConfigUpdate_DefaultTarget(t *testing.T) {
+	// Start an RLS server and set the throttler to always throttle requests.
+	rlsServer, _ := setupFakeRLSServer(t, nil)
+	overrideAdaptiveThrottler(t, alwaysThrottlingThrottler())
+
+	// Build RLS service config with a default target.
+	rlsConfig := buildBasicRLSConfigWithChildPolicy(t, t.Name(), rlsServer.Address)
+	backendCh1, backendAddress1 := startBackend(t)
+	rlsConfig.RouteLookupConfig.DefaultTarget = backendAddress1
+
+	// Register a manual resolver and push the RLS service config through it.
+	r := startManualResolverWithConfig(t, rlsConfig)
+
+	cc, err := grpc.Dial(r.Scheme()+":///", grpc.WithResolvers(r), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	defer cc.Close()
+
+	// Make an RPC and ensure it gets routed to the default target.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	makeTestRPCAndExpectItToReachBackend(ctx, t, cc, backendCh1)
+
+	// Change default_target field of the RLS config.
+	backendCh2, backendAddress2 := startBackend(t)
+	rlsConfig.RouteLookupConfig.DefaultTarget = backendAddress2
+
+	// Push the config update through the manual resolver.
+	scJSON, err := rlsConfig.ServiceConfigJSON()
+	if err != nil {
+		t.Fatal(err)
+	}
+	sc := internal.ParseServiceConfigForTesting.(func(string) *serviceconfig.ParseResult)(scJSON)
+	r.UpdateState(resolver.State{ServiceConfig: sc})
+
+	// The update is handled asynchronously by the LB policy. So, we don't know
+	// exactly when the default target will change. So, we keep retrying the RPC
+	// here until it gets routed to the new default policy or the context
+	// expires, at which point we fail the test.
+	done := grpcsync.NewEvent()
+	for {
```

Comment:
`for !done.HasFired() {` ?

Actually, why not make `done` a bool?

Reply:
Oh wait.  Why not just make this an infinite loop and `break <label>` instead of `done.Fire()`?

I guess we can't do that because of the `func`, but you could remove the `func` which appears to only exist so it can defer context cancelation, which could be done manually instead.